File: 22.括号生成.py
#
# @lc app=leetcode.cn id=22 lang=python3
#
# [22] 括号生成
#

# @lc code=start
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 0:
            return []
        ans = []
        # 注意这里用的是list来存储,也可以用string吧等下试试
        def backtrack(tmp: list, lindex: int, rindex: int):
            if len(tmp) == 2 * n:
                ans.append(''.join(tmp))
                return

            if lindex > 0:
                tmp.append('(')
                backtrack(tmp, lindex-1, rindex)
                # 因为这里修改了tmp结构，不影响这一次)的添加
                tmp.pop()
            if rindex > 0 and lindex < rindex:
                tmp.append(')')
                backtrack(tmp, lindex, rindex-1)
                tmp.pop()

        backtrack([], n, n)
        return ans
            

# @lc code=end

File: 22.括号生成.go
/*
 * @lc app=leetcode.cn id=22 lang=golang
 *
 * [22] 括号生成
 */

// @lc code=start
func generateParenthesis(n int) []string {
	if n == 0 {
		return []string{}
	}

	ans := []string{}

	backtrack(n, n, "", &ans)

	return ans
}

func backtrack(lindex int, rindex int, tmp string, ans *[]string) {
	if lindex == 0 && rindex == 0 {
		*ans = append(*ans, tmp)
		return
	}

	// 优先加(看看情况
	if lindex > 0 {
		backtrack(lindex-1, rindex, tmp+"(", ans)
	}

	// 左边符号一定多余右边符号,继续看加有括号的情况
	if rindex > 0 && lindex < rindex {
		backtrack(lindex, rindex-1, tmp+")", ans)
	}
}

// @lc code=end

